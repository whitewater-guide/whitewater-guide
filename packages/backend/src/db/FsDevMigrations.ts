// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
