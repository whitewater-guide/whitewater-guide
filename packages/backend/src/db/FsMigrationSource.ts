import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
