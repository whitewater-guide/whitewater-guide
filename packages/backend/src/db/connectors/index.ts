export * from './createConnectors';
export * from './OffsetConnector';
export * from './types';
