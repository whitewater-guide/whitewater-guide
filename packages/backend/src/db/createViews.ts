import { readFileSync } from 'fs';
import Knex from 'knex';
import path from 'path';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i--) {
    const file = path.resolve(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
