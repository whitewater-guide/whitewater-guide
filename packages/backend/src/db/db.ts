/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
