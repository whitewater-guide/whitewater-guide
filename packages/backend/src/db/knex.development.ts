import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

module.exports = development;
