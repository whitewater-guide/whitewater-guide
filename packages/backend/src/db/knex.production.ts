import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './dist/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './dist/seeds/production',
  },
  pool: {
    min: 0,
  },
  // debug: true,
} as any;

module.exports = development;
