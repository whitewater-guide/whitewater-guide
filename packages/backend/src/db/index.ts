import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './runSqlFile';
export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './types';
export * from './db';
export * from './queryBuilders';
