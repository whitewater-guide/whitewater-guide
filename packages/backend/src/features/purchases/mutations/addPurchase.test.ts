import db, { holdTransaction, rollbackTransaction } from '@db';
import { BOOM_USER_3500, BOOM_USER_3500_ID, TEST_USER } from '@seeds/01_users';
import {
  BOOM_PROMO_EU_CIS_ACTIVE,
  BOOM_PROMO_LATIN_REDEEMED,
  BOOM_PROMO_REGION_ACTIVE,
  BOOM_PROMO_REGION_REDEEMED,
} from '@seeds/12_boom_promos';
import { anonContext, fakeContext, runQuery, UUID_REGEX } from '@test';
import { ApolloErrorCodes, PurchaseInput, PurchasePlatform } from '@ww-commons';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const mutation = `
  mutation addPurchase($purchase: PurchaseInput!) {
    addPurchase(purchase: $purchase)
  }
`;

it('anon should fail', async () => {
  const purchase: PurchaseInput = {
    platform: PurchasePlatform.boomstarter,
    transactionId: BOOM_PROMO_REGION_ACTIVE,
    productId: 'region.georgia',
  };
  const result = await runQuery(mutation, { purchase }, anonContext());
  expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
});

it('duplicate purchase should return false when for same user', async () => {
  const purchase: PurchaseInput = {
    platform: PurchasePlatform.boomstarter,
    transactionId: BOOM_PROMO_REGION_REDEEMED,
    productId: 'region.norway', // different region from already redeemed, but should not matter
  };
  const result = await runQuery(mutation, { purchase }, fakeContext(TEST_USER));
  expect(result.errors).toBeUndefined();
  expect(result).toHaveProperty('data.addPurchase', false);
});

it('duplicate purchase should return throw when for different user', async () => {
  const purchase: PurchaseInput = {
    platform: PurchasePlatform.boomstarter,
    transactionId: BOOM_PROMO_REGION_REDEEMED,
    productId: 'region.norway', // different region from already redeemed, but should not matter
  };
  const result = await runQuery(mutation, { purchase }, fakeContext(BOOM_USER_3500));
  expect(result).toHaveGraphqlError(
    ApolloErrorCodes.MUTATION_NOT_ALLOWED,
    'Duplicate transaction',
  );
});

it('should throw on invalid input', async () => {
  const purchase = {
    platform: 'wrong',
    transactionId: 'x',
    productId: 'y',
  };
  const result = await runQuery(mutation, { purchase }, fakeContext(TEST_USER));
  expect(result).toHaveGraphqlValidationError();
});

describe('boomstarter', () => {
  it('should throw error for non-existing promo code', async () => {
    const purchase: PurchaseInput = {
      platform: PurchasePlatform.boomstarter,
      transactionId: 'xxxxxxxx',
      productId: 'region.norway',
    };
    const result = await runQuery(mutation, { purchase }, fakeContext(BOOM_USER_3500));
    expect(result).toHaveGraphqlError(
      ApolloErrorCodes.MUTATION_NOT_ALLOWED,
      'Bad promo code',
    );
  });

  it('should throw error when promo code and product do not match', async () => {
    const purchase: PurchaseInput = {
      platform: PurchasePlatform.boomstarter,
      transactionId: BOOM_PROMO_EU_CIS_ACTIVE,
      productId: 'region.norway',
    };
    const result = await runQuery(mutation, { purchase }, fakeContext(BOOM_USER_3500));
    expect(result).toHaveGraphqlError(
      ApolloErrorCodes.MUTATION_NOT_ALLOWED,
      'Promo code and product do not match',
    );
  });

  it('should throw error for used promo code', async () => {
    const purchase: PurchaseInput = {
      platform: PurchasePlatform.boomstarter,
      transactionId: BOOM_PROMO_LATIN_REDEEMED,
      productId: 'group.latin',
    };
    const result = await runQuery(mutation, { purchase }, fakeContext(BOOM_USER_3500));
    expect(result).toHaveGraphqlError(
      ApolloErrorCodes.MUTATION_NOT_ALLOWED,
      'Promo code already redeemed',
    );
  });

  describe('success', () => {
    let result: any;

    beforeEach(async () => {
      const purchase: PurchaseInput = {
        platform: PurchasePlatform.boomstarter,
        transactionId: BOOM_PROMO_EU_CIS_ACTIVE,
        productId: 'group.eu_cis',
      };
      result = await runQuery(mutation, { purchase }, fakeContext(BOOM_USER_3500));
    });

    it('should return true for correct promo code', () => {
      expect(result.errors).toBeUndefined();
      expect(result).toHaveProperty('data.addPurchase', true);
    });

    it('should save transaction', async () => {
      const transactions = await db().table('transactions').where({ transaction_id: BOOM_PROMO_EU_CIS_ACTIVE });
      expect(transactions).toHaveLength(1);
      expect(transactions[0]).toMatchObject({
        id: expect.stringMatching(UUID_REGEX),
        user_id: BOOM_USER_3500_ID,
        platform: PurchasePlatform.boomstarter,
        transaction_date: null,
        transaction_id: BOOM_PROMO_EU_CIS_ACTIVE,
        product_id: 'group.eu_cis',
        receipt: null,
        validated: true,
        extra: null,
      });
    });

    it('should mark promo code as redeemed', async () => {
      const promo = await db().table('boom_promos').where({ code: BOOM_PROMO_EU_CIS_ACTIVE }).first();
      expect(promo).toMatchObject({
        code: BOOM_PROMO_EU_CIS_ACTIVE,
        group_sku: 'group.eu_cis',
        redeemed: true,
      });
    });

  });

});
