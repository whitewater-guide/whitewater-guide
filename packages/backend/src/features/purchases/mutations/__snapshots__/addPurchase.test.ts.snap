// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw on invalid input 1`] = `
Object {
  "0": Object {
    "message": "\\"transactionId\\" length must be at least 3 characters long",
    "path": "purchase.transactionId",
  },
  "1": Object {
    "message": "\\"productId\\" length must be at least 3 characters long",
    "path": "purchase.productId",
  },
}
`;
